File: 卡码/代码随想录/1/11-图论/C++/graph/dfs/LeetCode_0797.cpp
//797. 所有可能的路径
#include <iostream>
#include <vector>
using namespace std;

class Solution {
	vector<vector<int>> res;
	vector<int> path;
public:
	void dfs(vector<vector<int>>& graph,int index){
		if(index==graph.size()-1){
			res.push_back(path);
			return;
		}
		for(int i=0;i<graph[index].size();i++){
			path.push_back(graph[index][i]);
			dfs(graph,graph[index][i]);
			path.pop_back();
		}
	}
	vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
		path.push_back(0);
		dfs(graph,0);
		return res;
	}
};

int main(){
  system("pause");
  return 0;
}

File: 卡码/代码随想录/Java/src/com/kurumi/codeThoughts/图论/dfs/LeetCode_0797.java
package com.kurumi.codeThoughts.图论.dfs;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

/**
 * @Author kurumi
 * @Date 2026-07-03 上午11:52:20
 * @Description 797. 所有可能的路径
 */
public class LeetCode_0797 {
    static class Solution {
        List<List<Integer>> res = new ArrayList<>();
        Deque<Integer> path = new ArrayDeque();

        public void dfs(int[][] graph, int index) {
            if (index == graph.length - 1) {
                res.add(new ArrayList<>(path));
                return;
            }
            for (int i = 0; i < graph[index].length; i++) {
                path.offer(graph[index][i]);
                dfs(graph, graph[index][i]);
                path.pollLast();
            }
        }

        public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
            path.add(0);
            dfs(graph, 0);
            return res;
        }
    }
}
